### PR TITLE
Fix nullable deleted column in API Keys table

### PR DESCRIPTION
**Pull request by davelopez**
By default, the `deleted` column is nullable even if it has a [default value of False](https://github.com/galaxyproject/galaxy/blob/1a984f27ab755b71a817f2bb95139af9fd43c7aa/lib/galaxy/model/__init__.py#L9700).

There is a chance, some external scripts that create API keys for users directly in the database might not set the `deleted` value explicitly causing possible incongruences with multiple API keys for the user not being explicitly deleted=False.

This restricts those cases by forcing a non-null value.

We could also try to handle possible null values in the API logic, I can change the approach if this is more desirable than altering the database schema at this point.

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

**Comment by mvdbeek** [comment]
Doesn't this potentially make old (and thus supposedly inactive) API keys valid ? Can we just fix those external scripts and call it a day?

**Comment by mvdbeek** [comment]
See also https://github.com/galaxyproject/galaxy/pull/14489 for context

File: lib/galaxy/model/migrations/alembic/versions_gxy/1d0564b6d7aa_make_api_keys_deleted_non_nullable.py
Comment (mvdbeek) [inline lib/galaxy/model/migrations/alembic/versions_gxy/1d0564b6d7aa_make_api_keys_deleted_non_nullable.py]: If you do want to alter these you'll have to make it so that only the newest API key is set to `deleted=False` and all older ones are set to `deleted=True`.

Comment (mvdbeek) [reply lib/galaxy/model/migrations/alembic/versions_gxy/1d0564b6d7aa_make_api_keys_deleted_non_nullable.py]: and of course only if there isn't one that is already valid for that user.

**Comment by davelopez** [comment]
> Doesn't this potentially make old (and thus supposedly inactive) API keys valid ?

If I remember correctly with https://github.com/galaxyproject/galaxy/pull/14489, only the **latest** API that is not deleted will be the valid one, so I guess the only problem will be with a **new** API key that is deleted=NULL

> Can we just fix those external scripts and call it a day?

I'm completely fine with that, ping @hexylena in case she has some opinions.

> Would you also need to add , nullable=True to https://github.com/galaxyproject/galaxy/blob/release_23.0/lib/galaxy/model/__init__.py#L9700 ?

Apparently, this is the default when you don't explicitly set the `nullable` property, so it won't change anything I guess.

File: lib/galaxy/model/migrations/alembic/versions_gxy/1d0564b6d7aa_make_api_keys_deleted_non_nullable.py
Comment (hexylena) [reply lib/galaxy/model/migrations/alembic/versions_gxy/1d0564b6d7aa_make_api_keys_deleted_non_nullable.py]:
```suggestion
    op.execute(table.update().where(table.c.deleted.is_(None)).values(deleted=True))
```

even without the sorting, safer to just have it fail closed.

**Comment by mvdbeek** [comment]
> Apparently, this is the default when you don't explicitly set the `nullable` property, so it won't change anything I guess.

I guess @nsoranzo meant that it needs to be changed to nullable=False, no ?

**Comment by nsoranzo** [comment]
> I guess @nsoranzo meant that it needs to be changed to nullable=False, no ?

Indeed, sorry!

**Comment by mvdbeek** [comment]
> only the **latest** API that is not deleted will be the valid one, so I guess the only problem will be with a **new** API key that is deleted=NULL

right ... so if an admin set deleted=False on the latest API key your old key becomes active ... it just seems like there is too much room for error here.

**Comment by mvdbeek** [comment]
what happens when you declare a column as non-nullable in the migration but you have null values in the database ? if that only blocks inserting new null values then that would also be a valid option.

**Comment by davelopez** [comment]
> what happens when you declare a column as non-nullable in the migration but you have null values in the database ?

The migration will fail with:
```
sqlalchemy.exc.IntegrityError: (psycopg2.errors.NotNullViolation) column "deleted" of relation "api_keys" contains null values
```

Agree, I will change the delete=NULL existing rows to delete=True as suggested, feels much safer, in the worst case the user will need to recreate the API key because the one that may be used was discarded in the migration.

**Comment by bgruening** [comment]
This probably impacts only EU. If it's all easier, then let's fix this on dev and don't introduce a migration in a stable branch.

File: lib/galaxy/model/migrations/alembic/versions_gxy/1d0564b6d7aa_make_api_keys_deleted_non_nullable.py
Comment (davelopez) [reply lib/galaxy/model/migrations/alembic/versions_gxy/1d0564b6d7aa_make_api_keys_deleted_non_nullable.py]: Done in the rebase to keep the migration script in a single commit, thanks!

**Comment by hexylena** [comment]
> > Can we just fix those external scripts and call it a day?
>
> I'm completely fine with that, ping @hexylena in case she has some opinions.

I mean, yea, we can. We are doing that separately. That said, this column allowed nulls which semantically do not make sense, and we should fix that underlying issue as well at some point, how that happens (23.0 vs dev) all is fine for me. But now that it 'fails closed', this looks better to me.

**Review by mvdbeek — APPROVED** [review]
Looks good to me, @jdavcs can you also take a look ?

**Comment by jdavcs** [comment]
Both alter statements (upgrade and downgrade) should be wrapped in a batch operations context manager:
```py
with op.batch_alter_table(table_name) as batch_op:
    batch_op.alter_column(column_name, existing_type=sa.Boolean(), nullable=True, server_default=None)
```
Note that we *do not* need to pass `table name` to `alter_column`, and we call it on `batch_op`, not `op`.

The reason for that is that SQLite has very limited support for ALTER statements. The batch mode uses the ["move and copy" approach](https://alembic.sqlalchemy.org/en/latest/batch.html).

The rest looks fine.

**Comment by jdavcs** [comment]
Also, not a dealbreaker, but would it be possible to merge this *after* #15811? Otherwise, it would have to be edited after that PR is merged. With 15811, this revision script would look like this:
```py
with transaction():
   op.execute(...)
   alter_column(...)
```
In addition, this revision would be tested (that PR adds a test to the CI so that all revisions go through a downgrade>upgrade test sequence on both postgresql *and* sqlite).

**Comment by nsoranzo** [comment]
> Also, not a dealbreaker, but would it be possible to merge this _after_ #15811? Otherwise, it would have to be edited after that PR is merged. With 15811, this revision script would look like this:

This seems another reason to target dev instead of 23.0 ?

**Comment by davelopez** [comment]
Rebased and retargeted to `dev`.
@jdavcs a final review to see if I included your requested changes properly would be much appreciated :beers:

**Comment by jdavcs** [comment]
@davelopez  I've added the utility in #16009. With that, instead of  `with op.batch...`  you can just call `alter_column(` passing the table name + the same arguments you passed to the batch op function.

File: lib/galaxy/model/migrations/alembic/versions_gxy/b855b714e8b8_make_api_keys_deleted_non_nullable.py
Comment (nsoranzo) [inline lib/galaxy/model/migrations/alembic/versions_gxy/b855b714e8b8_make_api_keys_deleted_non_nullable.py:39]:
```suggestion
    alter_column(table_name, column_name, existing_type=sa.Boolean(), nullable=True, existing_server_default=False)
```

Comment (davelopez) [reply lib/galaxy/model/migrations/alembic/versions_gxy/b855b714e8b8_make_api_keys_deleted_non_nullable.py:39]: This will modify the original column's behavior after a downgrade, right?
With this proposed change, after the `downgrade` any new "deleted=null" values will be defaulted to "deleted=False". I don't think this is what we want unless I'm missing something :thinking:

Comment (nsoranzo) [reply lib/galaxy/model/migrations/alembic/versions_gxy/b855b714e8b8_make_api_keys_deleted_non_nullable.py:39]: `False` is the current and future default for the `deleted` column.

Comment (nsoranzo) [reply lib/galaxy/model/migrations/alembic/versions_gxy/b855b714e8b8_make_api_keys_deleted_non_nullable.py:39]: I was mistaken here, we set a `default` but not a `server_default` for this column (actually for any column). Thanks to David, I discovered `server_default` is the one that should be used if you want the default to apply also outside of SQLAlchemy (i.e. when using SQL statements).

File: lib/galaxy/model/__init__.py
Comment (nsoranzo) [inline lib/galaxy/model/__init__.py]:
```suggestion
    deleted = Column(Boolean, index=True, server_default=False, nullable=False)
```

File: lib/galaxy/model/migrations/alembic/versions_gxy/b855b714e8b8_make_api_keys_deleted_non_nullable.py
Comment (jdavcs) [reply lib/galaxy/model/migrations/alembic/versions_gxy/b855b714e8b8_make_api_keys_deleted_non_nullable.py:39]: `False` is the default in Alembic's `alter_column` function. However, in SQLAlchemy, the default is `None` ([Column costructor](https://docs.sqlalchemy.org/en/20/core/metadata.html#sqlalchemy.schema.Column.__init__)). Thus, if we run this migration, Alembic's default is applied, but if we skip migrations (e.g. new installation), it'll be SQLAlchemy's default. Therefore, given that we want the downgrade function to reverse the upgrade, we want the database to have `null` as the default value, which (in the case of postgres) is done by dropping the default (not setting it to false). At least that's my understanding.

Comment (jdavcs) [reply lib/galaxy/model/migrations/alembic/versions_gxy/b855b714e8b8_make_api_keys_deleted_non_nullable.py:39]: More on this :point_up: Essentially, even though the resulted value in the database will be the same, `default` is not the same as `server_default` (I'm looking at [these docs](https://docs.sqlalchemy.org/en/20/core/defaults.html)). Since we do not use `server_default` in the model's definition (and why would we), I think it would make sense to ensure the column DDL in the database does not change as a result of running an upgrade->downgrade sequence?

**Review by jdavcs — APPROVED** [review]

**Comment by jdavcs** [comment]
Test failures: `server_default` cannot be a bool. None or str. https://docs.sqlalchemy.org/en/20/core/metadata.html#sqlalchemy.schema.Column.params.server_default

**Review by nsoranzo — APPROVED** [review]

**Comment by jdavcs** [comment]
Thank you, @davelopez and everyone for such a thorough review!

**Comment by mvdbeek** [comment]
this definitely needs to go into the release announcement, since it'll invalidate old api keys

**Pull request merged**
